### PR TITLE
Improved performance

### PR DESCRIPTION
**Pull request by Spadavecchia**

### CODE REVIEW

File: lib/liquid/combinators/tags/capture.ex [hunk 1]
```diff
@@ -38,4 +38,22 @@ defmodule Liquid.Combinators.Tags.Capture do
       fn combinator -> optional(combinator, parsec(:__parse__) |> tag(:parts)) end
     )
   end
+
+  def tag2 do
```
Comment (Eptis) [inline lib/liquid/combinators/tags/capture.ex]: Why is this function named like this?

File: lib/liquid/ast.ex [hunk 2]
```diff
@@ -0,0 +1,45 @@
+defmodule Liquid.Ast do
```
Comment (Eptis) [inline lib/liquid/ast.ex]: Can you add module documentation?

File: lib/liquid/combinators/tags/end_block.ex [hunk 3]
```diff
@@ -0,0 +1,27 @@
+defmodule Liquid.Combinators.Tags.EndBlock do
+  @moduledoc """
+  Verifies when block is closed and send the AST to end the block
+  ```
```
Comment (Eptis) [inline lib/liquid/combinators/tags/end_block.ex]: Why are these here?

File: lib/liquid/tokenizer.ex [hunk 4]
```diff
@@ -0,0 +1,30 @@
+defmodule Liquid.Tokenizer do
+  @moduledoc """
+  Prepares markup to be parsed
```
Comment (Eptis) [inline lib/liquid/tokenizer.ex]: Can you add more documentation on _how_? For instance: Prepares markup to be parsed by splitting up...

File: lib/liquid/combinators/tags/capture.ex (hunk 1 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/capture.ex]: We want preserve the old tag until we finished doing the improve of performance. Once the proposal is reviewed and accepted, we're going to delete the old tag and rename tag2 -> tag

File: lib/liquid/ast.ex (hunk 2 above)
Comment (Spadavecchia) [reply lib/liquid/ast.ex]: Sure, we'll add documentation in the next commits

File: lib/liquid/combinators/tags/end_block.ex (hunk 3 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/end_block.ex]: This module generalize the close of all blocks and it marks the end of internal recursion.

File: lib/liquid/tokenizer.ex (hunk 4 above)
Comment (Spadavecchia) [reply lib/liquid/tokenizer.ex]: Sure, we'll add more documentation in the next commit

File: lib/liquid/combinators/tags/capture.ex (hunk 1 above)
Comment (Eptis) [reply lib/liquid/combinators/tags/capture.ex]: Shouldn't that change be part of accepting this PR?

File: lib/liquid/combinators/tags/end_block.ex (hunk 3 above)
Comment (Eptis) [reply lib/liquid/combinators/tags/end_block.ex]: No I mean why are the "```" in the documentation?

File: lib/liquid/combinators/tags/capture.ex (hunk 1 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/capture.ex]: changing it should break some tests which are dependents of reused code

File: lib/liquid/combinators/tag.ex [hunk 5]
```diff
@@ -59,4 +81,32 @@ defmodule Liquid.Combinators.Tag do
     |> ignore(string("end" <> tag_name))
     |> parsec(:end_tag)
   end
+
+  def store_tag_in_context(_rest, [{:block, tag}] = acc, %{tags: tags} = context, _line, _offset) do
```
[the remark this answers is not in the document]
Comment (Eptis) [reply lib/liquid/combinators/tag.ex]: and also `@spec`s?

File: lib/liquid/combinators/tag.ex [hunk 6]
```diff
@@ -59,4 +81,32 @@ defmodule Liquid.Combinators.Tag do
...
+    tag_name = tag |> Keyword.keys() |> hd() |> to_string()
+    {acc, %{context | tags: [tag_name | tags]}}
+  end
+
+  def check_allowed_tags(_, acc, %{tags: []} = context, _, _, _) do
```
Comment (Eptis) [inline lib/liquid/combinators/tag.ex]: Could you add `@doc`umentation and `@spec`s?

File: lib/liquid/combinators/tags/custom_tag.ex [hunk 7]
```diff
@@ -23,16 +23,33 @@ defmodule Liquid.Combinators.Tags.CustomTag do
...
+    |> optional(markup())
     |> parsec(:end_tag)
-    |> tag(:custom_tag)
+    |> traverse({__MODULE__, :check_customs, []})
   end
 
-  defp name do
-    valid_name()
-    |> unwrap_and_tag(:custom_name)
+  def check_customs(_, [params | tag], %{tags: tags} = context, _, _) do
```
Comment (Eptis) [inline lib/liquid/combinators/tags/custom_tag.ex]: Could you add `@doc`umentation and `@spec`s? What does it check for?

File: lib/liquid/combinators/tags/custom_tag.ex [hunk 8]
```diff
@@ -23,16 +23,33 @@ defmodule Liquid.Combinators.Tags.CustomTag do
...
+    [tag_name] = tag
+    name = String.to_atom(tag_name)
```
Comment (Eptis) [inline lib/liquid/combinators/tags/custom_tag.ex]: Please use `String.to_existing_atom/1`, the tag should already be registered with an atom.

File: lib/liquid/nimble_translator.ex [hunk 9]
```diff
@@ -64,10 +62,11 @@ defmodule Liquid.NimbleTranslator do
...
         :for -> For.translate(markup)
-        :if -> If.translate(markup)
-        :unless -> Unless.translate(markup)
-        :elsif -> If.translate(markup)
-        :else -> process_node(markup)
+        :if -> If.translate(:if, markup)
+        :unless -> If.translate(:unless, markup)
+        :elsif -> If.translate(:if, markup)
+        :else -> [body: body_parts] = markup
+        process_node(body_parts)
```
Comment (Eptis) [inline lib/liquid/nimble_translator.ex]: Can you indent?

File: lib/liquid/combinators/tag.ex [hunk 10]
```diff
@@ -38,17 +40,40 @@ defmodule Liquid.Combinators.Tag do
...
+    |> parsec(:start_tag)
+    |> ignore(string(tag_name))
+    |> combinator.()
+    |> parsec(:end_tag)
+    |> tag(String.to_atom(tag_name))
+    |> tag(:sub_block)
+    |> traverse({__MODULE__, :check_allowed_tags, [allowed_tags]})
+  end
+
+  def define_block(tag_name, combinator_head \\ & &1, separator \\ " ")
```
Comment (basdirks) [inline lib/liquid/combinators/tag.ex]: Could you add `@doc` and `@spec`?

File: lib/liquid/combinators/tag.ex [hunk 11]
```diff
@@ -38,17 +40,40 @@ defmodule Liquid.Combinators.Tag do
...
+
+  def define_block(tag_name, combinator_head, separator) do
+    tag_name
+    |> open_tag(combinator_head, separator)
+    |> tag(String.to_atom(tag_name))
+    |> tag(:block)
+    |> traverse({__MODULE__, :store_tag_in_context, []})
+  end
+
   def define_inverse_open(tag_name, combinator_head \\ & &1) do
```
Comment (basdirks) [inline lib/liquid/combinators/tag.ex]: Could you add `@doc` and `@spec`?

File: lib/liquid/combinators/tag.ex [hunk 12]
```diff
@@ -38,17 +40,40 @@ defmodule Liquid.Combinators.Tag do
...
     tag_name
     |> open_tag(combinator_head)
     |> optional(parsec(:__parse__))
     |> tag(String.to_atom(tag_name))
   end
 
-  def open_tag(tag_name, combinator \\ & &1) do
+  def open_tag(tag_name, combinator \\ & &1, separator \\ " ")
```
Comment (basdirks) [inline lib/liquid/combinators/tag.ex]: Could you add `@doc` and `@spec`?

File: lib/liquid/ast.ex [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  Literals (any markup which is not liquid variable or tag) are slow to be processed by Nimble thus
+  this module improve performance between 30% and 100% depending how much text is processed.
+  """
+  alias Liquid.{Tokenizer, Parser, Block}
+
+  @doc """
+  Recursively builds the AST taking a markup, or a tuple with a literal and a rest markup.
+  It uses `context` to validate the correct opening and close of blocks and sub blocks.
+  """
+  @spec build(String.t() | {String.t(), String.t()}, Keyword.t(), List.t()) ::
```
Comment (basdirks) [inline lib/liquid/ast.ex]: Can you go with a single style of type, preferably shorthands? In this case:

```
@spec build(binary | {binary, binary}, Keyword.t(), list) :: {:ok, list, Keyword.t(), binary} | {:error, binary, binary}
```

Thanks :)

File: lib/liquid/combinators/tag.ex [hunk 14]
```diff
@@ -59,4 +84,32 @@ defmodule Liquid.Combinators.Tag do
     |> ignore(string("end" <> tag_name))
     |> parsec(:end_tag)
   end
+
+  def store_tag_in_context(_rest, [{:block, tag}] = acc, %{tags: tags} = context, _line, _offset) do
+    tag_name = tag |> Keyword.keys() |> hd() |> to_string()
+    {acc, %{context | tags: [tag_name | tags]}}
+  end
+
+  def check_allowed_tags(_, acc, %{tags: []} = context, _, _, _) do
```
Comment (basdirks) [inline lib/liquid/combinators/tag.ex]: What are the first, fourth, and fifth arguments for?

File: lib/liquid/combinators/tags/custom_tag.ex (hunk 8 above)
Comment (Spadavecchia) [reply lib/liquid/combinators/tags/custom_tag.ex]: This runs in compilation and knows nothing about the register.

File: lib/liquid/combinators/tag.ex (hunk 14 above)
Comment (fabianherrera) [reply lib/liquid/combinators/tag.ex]: Those are coming from Nimble parser library traverse sentence, this represent:
(_res, = rest of markup to be parser, args = parser accumulator at point traverse was called, context = a map with specific parameters you wish to create to store some interesting content, _line = line number where excecution is, _offset =  internal offset counter, _last = any_custom_parameters_send_via_traverse_to_internal_function.

https://hexdocs.pm/nimble_parsec/NimbleParsec.html#traverse/3

We are including this names in orer to complement this parameters names.

File: lib/liquid/ast.ex (hunk 13 above)
Comment (fabianherrera) [reply lib/liquid/ast.ex]: Done

File: lib/liquid/combinators/tag.ex (hunk 12 above)
Comment (fabianherrera) [reply lib/liquid/combinators/tag.ex]: Done

File: lib/liquid/combinators/tag.ex (hunk 10 above)
Comment (fabianherrera) [reply lib/liquid/combinators/tag.ex]: Done.

File: lib/liquid/combinators/tag.ex (hunk 6 above)
Comment (fabianherrera) [reply lib/liquid/combinators/tag.ex]: Done

File: lib/liquid/combinators/tag.ex [hunk 15]
```diff
@@ -24,9 +24,10 @@ defmodule Liquid.Combinators.Tag do
       MyParser.ignorable("{% ignorable T12 %}")
       #=> {:ok, {:ignorable, [12]}, "", %{}, {1, 0}, 2}
   """
-  def define_closed(tag_name, combinator_head \\ & &1, combinator_body \\ & &1) do
+  def define_closed(tag_name, combinator_head \\ & &1, combinator_body \\ & &1, separator \\ " ")
```
[the remark this answers is not in the document]
Comment (fabianherrera) [reply lib/liquid/combinators/tag.ex]: Done

File: lib/liquid/combinators/tag.ex [hunk 16]
```diff
@@ -38,17 +39,38 @@ defmodule Liquid.Combinators.Tag do
     |> tag(String.to_atom(tag_name))
   end
 
+  def define_sub_block(tag_name, allowed_tags, combinator \\ & &1) do
```
[the remark this answers is not in the document]
Comment (fabianherrera) [reply lib/liquid/combinators/tag.ex]: Done

File: lib/liquid/combinators/general.ex [hunk 17]
```diff
@@ -387,5 +387,25 @@ defmodule Liquid.Combinators.General do
...
     ])
     |> optional(times(parsec(:logical_condition), min: 1))
+    |> tag(:conditions)
+  end
+
+  @doc """
+  Parses a `Liquid` Custom tag's name, isolates custom tag name from markup.
+  """
+  @spec valid_tag_name() :: NimbleParsec.t()
+  def valid_tag_name do
```
[the remark this answers is not in the document]
Comment (fabianherrera) [reply lib/liquid/combinators/general.ex]: Done

[the remark this answers is not in the document]
Comment (Eptis) [reply lib/liquid/combinators/general.ex]: But why is this function named `valid_tag_name`?

File: lib/liquid/combinators/tags/custom_tag.ex (hunk 8 above)
Comment (Eptis) [reply lib/liquid/combinators/tags/custom_tag.ex]: Then there is no problem with using `String.to_existing_atom/1` right?
